projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/exception.py

```diff
@@ -0,0 +1,184 @@
+# Copyright 2021-2023 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+from typing import Any
+from typing import Optional
+
+from vdk.internal.core.errors import BaseVdkError
+from vdk.internal.core.errors import ResolvableBy
+
+
+class IngestionException(BaseVdkError):
+    """
+    Base Exception for all custom exceptions related to the ingestion process.
+    This is intended to catch general exceptions that do not fit into more specific categories.
+    """
+
+    def __init__(self, message: str, vdk_type: ResolvableBy = None):
+        super().__init__(None, vdk_type, message)
+
+
+class IngestionConfigurationException(IngestionException):
+    """
+    Raised when there are configuration-related issues during the ingestion process.
+    """
+
+    def __init__(self, message: str, vdk_type: ResolvableBy = None):
+        super().__init__(message=message, vdk_type=vdk_type)
+
+
+class PayloadIngestionException(IngestionException):
+    """
+    Base Exception for all payload-related issues during the ingestion process.
+    """
+
+    def __init__(
+        self,
+        payload_id: str,
+        message: str,
+        destination_table: str = "",
+        target: str = "",
+        vdk_type: ResolvableBy = None,
+    ):
+        """
+        :param payload_id: ID of the payload. This is a way for user to find out which payload failed to ingest.
+                           Left empty if such information is not available
+        """
+        super().__init__(message=message, vdk_type=vdk_type)
```

Comment:
Another thing I've been wondering. 

Do I want to pass vdk_type in each exception class ? 
Or do I just inject it dynamically in report_and_throw 

But not all exception are raised using report_and_throw perhaps.

The way ingestion works so far (and I haven't changed that), report_and_throw is used for exception at the aggregate level that are called on the main thread (e.g close()) but for code running in background threads , report_and_throw is not used (this avoids issues with concurrent access over globals in report_and_throw)

Reply:
We don't use `report_and_throw` for non-vdk exceptions. You probably won't find a code snippet like

```
report_and_throw(new ValueError("My message, whatever"))
```

> But not all exception are raised using report_and_throw perhaps.

This is correct.

All in all, it's not a bad idea to pass the type in both cases.